pkg/routeagent_driver/handlers/ovn/submarinerGWRouteHandler.go

```diff
@@ -0,0 +1,206 @@
+type GatewayRouteHandler struct {
+	event.HandlerBase
+	mutex           sync.Mutex
+	smClient        submarinerClientset.Interface
+	config          *environment.Specification
+	remoteEndpoints map[string]*submarinerv1.Endpoint
+	isGateway       bool
+	nextHopIP       string
+}
+
+func NewGatewayRouteHandler(env *environment.Specification, smClientSet submarinerClientset.Interface) *GatewayRouteHandler {
+	// We'll panic if env is nil, this is intentional
+	return &GatewayRouteHandler{
+		config:          env,
+		smClient:        smClientSet,
+		remoteEndpoints: map[string]*submarinerv1.Endpoint{},
+	}
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) Init() error {
+	logger.Infof("Starting GatewayRouteHandler")
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) GetName() string {
+	return "submariner-gw-route-handler"
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) GetNetworkPlugins() []string {
+	// TODO enable when we switch to new implementation
+	// return []string{cni.OVNKubernetes}
+	return []string{}
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) LocalEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	nextHopIP, err := gatewayRouteHandler.getNextHopOnK8sMgmtIntf(endpoint)
+
+	if err != nil || nextHopIP == nil {
+		return errors.Wrapf(err, "error getting the ovn kubernetes management interface IP")
+	}
+
+	gatewayRouteHandler.nextHopIP = nextHopIP.String()
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	if gatewayRouteHandler.isGateway {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			_, err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+				gatewayRouteHandler.getGatewayRoute(endpoint), metav1.CreateOptions{})
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrapf(err, "error processing the remote endpoint creation for %q", endpoint.Name)
+			}
+			return nil
+		})
+
+		return errors.Wrapf(retryErr, "processing the remote endpoint event failed even after retry")
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) RemoteEndpointRemoved(endpoint *submarinerv1.Endpoint) error {
+	gatewayRouteHandler.mutex.Lock()
+	defer gatewayRouteHandler.mutex.Unlock()
+
+	delete(gatewayRouteHandler.remoteEndpoints, endpoint.Name)
+
+	if gatewayRouteHandler.isGateway {
+		if err := gatewayRouteHandler.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Delete(context.TODO(),
+			endpoint.Name, metav1.DeleteOptions{}); err != nil {
+			return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (gatewayRouteHandler *GatewayRouteHandler) getNextHopOnK8sMgmtIntf(*submarinerv1.Endpoint) (*net.IP, error) {
```

Comment:
Can we move this function to some util/other ovn file?
It just takes local cluster cidrs, I can see that we already have similar function in https://github.com/submariner-io/submariner/blob/devel/pkg/routeagent_driver/handlers/ovn/host_networking.go#L113C21-L113C44

Reply:
Done.